keps/sig-storage/3301-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,681 @@
+
+
+### Goals
+
+- Make `CSINode.Spec.Drivers[*].Allocatable.Count` mutable.
+- Enable CSI drivers to define the interval at which the `Allocatable.Count` value on each node is updated through the `CSIDriver` object.
+- Automatically update `CSINode.Spec.Drivers[*].Allocatable.Count` upon detecting a failure in volume attachment due to insufficient capacity.
+
+
+### Non-Goals
+
+- Modifying the core scheduling logic of Kubernetes.
+- Implementing cloud provider-specific solutions within Kubernetes core.
+- Re-scheduling pods stuck in a `ContainerCreating` state.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I want the reported attachment capacity on nodes to accurately reflect the actual capacity, so that
+stateful pods are reliably scheduled and do not become stuck in a `ContainerCreating` state due to insufficient capacity.
+
+#### Story 2
+
+As a cluster operator, I use volumes during node setup for components like kubelet, containerd, and additional drivers. These
+boot volumes, which are not managed by CSI, may be detached after setup, and I need a way to reclaim these slots for other
+uses. The current static capacity reporting doesn't allow for this flexibility.
+
+#### Story 3
+
+As a cluster operator, I need the Kubernetes scheduler to accurately count the number of available device slots for both storage volumes and network interfaces.
+On certain EC2 instance types, network interfaces and volumes share device slots, and network interfaces may be dynamically
+attached after the CSI driver is registered. This results in an inaccurate `Allocatable.Count` for volumes, causing stateful
+pods to be scheduled on nodes with insufficient capacity, ultimately getting stuck in a `ContainerCreating` state.
+
+### Risks and Mitigations
+
+The following risks are identified:
+
+- Frequent updates/retrieval of the `CSINode` object
+could increase API server load.
+
+- Frequent calls to a CSI driver's `NodeGetInfo` RPC
+endpoint may become expensive, particularly if the
+operation involves retrieving information from a
+remote server or performing resource-intensive tasks.
+Specifically, this is a concern at scale, where the
+cumulative cost of multiple nodes repeatedly
+querying for updates is more impactful.
+
+The risks are mitigated as follows:
+
+- The use of the Kubernetes informer pattern in the scheduler. The
+scheduler uses a `CSINode` informer and lister to
+efficiently access and watch `CSINode` objects.
+
+- Allow users to opt in to this
+feature at a per-CSI driver granularity. Administrators will
+also be able to fine-tune the interval update value as
+per their specific requirement.
+
+
+## Design Details
+
+### Feature Gate
+
+A new feature gate - `MutableCSINode` - will be introduced to control the functionality implemented by this KEP. When the
+feature gate is disabled, the `CSINode` object will remain immutable, maintaining the current behavior.
+
+### API Changes
+
+#### CSINode
```

Comment:
This involves API change.  @msau42 is an API approver.  Please make her an approver.